tests/queries/tests.py

```diff
@@ -4602,3 +4602,74 @@ def test_ticket_23622(self):
+            list(NamedCategory.objects.comment("").comment("some comment"))
+        self.assertIn("SELECT /* ! STRAIGHT_JOIN */ ", captured_queries[0]["sql"])
+        self.assertIn("SELECT /*  */ /* some comment */ ", captured_queries[1]["sql"])
+
+    def test_select_count(self):
+        with CaptureQueriesContext(connection) as captured_queries:
+            NamedCategory.objects.comment("COUNT").count()
+        self.assertIn("SELECT /* COUNT */ COUNT", captured_queries[0]["sql"])
+
+    def test_comment_before_distinct(self):
+        with CaptureQueriesContext(connection) as captured_queries:
+            list(NamedCategory.objects.comment("DISTINCT").all().distinct())
+        self.assertIn("SELECT /* DISTINCT */ DISTINCT", captured_queries[0]["sql"])
+
+    def test_select_subquery(self):
+        with CaptureQueriesContext(connection) as captured_queries:
+            list(
+                NamedCategory.objects.annotate(
+                    foo=(
+                        Tag.objects.filter(category=OuterRef("id"))
+                        .comment("a subquery comment")
+                        .values("name")[:1]
+                    )
+                ).comment("a comment")
+            )
+        sql = captured_queries[0]["sql"]
+        self.assertIn("(SELECT /* a subquery comment */ ", sql)
+        self.assertIn("SELECT /* a comment */ ", sql)
+
+    def test_select_aggregate(self):
+        with CaptureQueriesContext(connection) as captured_queries:
+            Tag.objects.comment("a subquery comment").values("parent").annotate(
+                tag_per_parent=Count("pk")
+            ).aggregate(Max("tag_per_parent")),
+        sql = captured_queries[0]["sql"]
+        self.assertIn("(SELECT /* a subquery comment */ ", sql)
+
+    def test_update(self):
+        with CaptureQueriesContext(connection) as captured_queries:
+            NamedCategory.objects.comment("this is an update").update(name="monty")
+        self.assertIn("UPDATE /* this is an update */ ", captured_queries[0]["sql"])
+
+    def test_stops_delimiters_in_message(self):
+        for comment in ["foo=/a/b/c*/", "/*foo=/a/b/c", "**//SELECT nothing;//**"]:
+            msg = (
+                "Cannot pass strings containing /* or */ to comment(). "
+                "Escape or strip these delimiters before calling comment()."
+            )
+
+            with self.subTest(comment), self.assertRaisesMessage(ValueError, msg):
+                NamedCategory.objects.comment(comment)
+
+    def test_delete(self):
+        with CaptureQueriesContext(connection) as captured_queries:
+            NamedCategory.objects.create(name="test")
+            NamedCategory.objects.comment("DELETE 1").all().delete()
+
+            NamedCategory.objects.create(name="test")
+            NamedCategory.objects.comment("DELETE 2").all().delete()
+        self.assertIn("DELETE FROM /* DELETE 1 */", captured_queries[1]["sql"])
```

Comment:
1. We don't need to create any data in order to run `DELETE`
2. Only one delete query needed
3. Use `startswith` and more realistic example data

```suggestion
            NamedCategory.objects.comment("blog/views.py:50").all().delete()
        self.assertTrue(
            captured_queries[0]["sql"].startswith(
                "DELETE /* blog/views.py:50 */ FROM", 
            )
        )
```

Reply:
The problem here is that delete does not generate delete query, but select:

This

```
NamedCategory.objects.comment("DELETE 2").delete()
```

will fail:
```
FAIL: test_delete (queries.tests.SqlCommentsInQueriesTests.test_delete)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/viciu/prog/django/tests/queries/tests.py", line 4670, in test_delete
    self.assertIn("DELETE FROM /* DELETE 1 */", [q['sql'] for q in captured_queries])
AssertionError: 'DELETE FROM /* DELETE 1 */' not found in 
['SELECT /* DELETE 2 */ "queries_dumbcategory"."id", "queries_namedcategory"."dumbcategory_ptr_id", "queries_namedcategory"."name" FROM "queries_namedcategory" INNER JOIN "queries_dumbcategory" ON ("queries_namedcategory"."dumbcategory_ptr_id" = "queries_dumbcategory"."id")']

```


therefore I was experimenting with adding some records first and then deleting them. If I understand correctly it looks like django fetches records to delete first, then updates related records with nulls, then finally deletes them one by one (including related records):

```
FAIL: test_delete (queries.tests.SqlCommentsInQueriesTests.test_delete)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/viciu/prog/django/tests/queries/tests.py", line 4670, in test_delete
    self.assertIn("DELETE FROM /* DELETE 1 */", [q['sql'] for q in captured_queries])
AssertionError: 'DELETE FROM /* DELETE 1 */' not found in 
[
'INSERT INTO "queries_dumbcategory" ("id") VALUES (NULL)', 
'INSERT INTO "queries_namedcategory" ("dumbcategory_ptr_id", "name") VALUES (1, \'test\')', 
'SELECT /* DELETE 1 */ "queries_dumbcategory"."id", "queries_namedcategory"."dumbcategory_ptr_id", "queries_namedcategory"."name" FROM "queries_namedcategory" INNER JOIN "queries_dumbcategory" ON ("queries_namedcategory"."dumbcategory_ptr_id" = "queries_dumbcategory"."id")', 
'UPDATE "queries_tag" SET "category_id" = NULL WHERE "queries_tag"."category_id" IN (1)', 
'DELETE FROM "queries_namedcategory" WHERE "queries_namedcategory"."dumbcategory_ptr_id" IN (1)',
'DELETE FROM "queries_dumbcategory" WHERE "queries_dumbcategory"."id" IN (1)', 

'INSERT INTO "queries_dumbcategory" ("id") VALUES (NULL)', 
'INSERT INTO "queries_namedcategory" ("dumbcategory_ptr_id", "name") VALUES (2, \'test\')', 
'SELECT /* DELETE 2 */ "queries_dumbcategory"."id", "queries_namedcategory"."dumbcategory_ptr_id", "queries_namedcategory"."name" FROM "queries_namedcategory" INNER JOIN "queries_dumbcategory" ON ("queries_namedcategory"."dumbcategory_ptr_id" = "queries_dumbcategory"."id")', 
'UPDATE "queries_tag" SET "category_id" = NULL WHERE "queries_tag"."category_id" IN (2)', 
'DELETE FROM "queries_namedcategory" WHERE "queries_namedcategory"."dumbcategory_ptr_id" IN (2)', 
'DELETE FROM "queries_dumbcategory" WHERE "queries_dumbcategory"."id" IN (2)']
```